hail/src/main/scala/is/hail/expr/ir/lowering/LoweringPipeline.scala

```diff
@@ -0,0 +1,33 @@
+package is.hail.expr.ir.lowering
+
+import is.hail.expr.ir.{BaseIR, ExecuteContext, Optimize}
+import is.hail.utils._
+
+case class LoweringPipeline(lowerings: IndexedSeq[LoweringPass]) {
```

Comment:
I'm not sure only capturing the sequence of lowerings, with just a boolean `optimize` flag, is general enough. I think a pipeline should be a sequence of lowerings and optimizations, where some optimizations may be specific to a certain stage of lowering.

Maybe there could be `LoweringPass` and `OptimizerPass`, which are both subtraits of `CompilerPass`, where an `OptimizerPass` starts and ends in the same `IRState`, and a `CompilerPipeline` is a sequence of `CompilerPass`es.

For example, it might make sense to split up `Simplify`, and wait to run table simplifications until after matrixtable to table lowering.

Reply:
Yeah, this sounds right. This PR is a rough first cut at some structured control flow in the compiler, so doing your refactor will be quite easy when it goes in.